New Meshline

Mugen87 opened the pull request:
This PR introduces a new and flexible way of drawing lines via [Screen-Space Projected Lines](https://mattdesl.svbtle.com/drawing-lines-is-hard#screenspace-projected-lines_2). The implementation is based on https://github.com/spite/THREE.MeshLine by @spite.

[**Demo**](https://yume.human-interactive.org/examples/meshline/webgl_meshline.html)

You can create a mesh line like this:

```javascript
var line = new THREE.MeshLine( geometry, material );
```
- `geometry` can be a `Geometry` or `BufferGeometry`. Users can provide vertices (position) and color data (for vertex colors).
- `material` must be a `MeshBasicMaterial`. It's the only material that is supported right now. You can use three new material properties to control the visual appearance of the line: `meshLine`(like `morphTargets`) , `meshLineWidth`, `meshLineSizeAttenuation`. Other material properties like `color`, `transparent` or `opacity` also work.
- Besides, there is a third and optional parameter of `MeshLine` that can be used to create nice curve shapes (`widthCallback`).

As soon as we commit to a final API design and naming conventions, i will also create some docs with a subsequent PR.

There is still an open issue. I'm not sure how to handle the situation if the geometry of a mesh line changes. `MeshLine` always needs to process the given geometry to prepare the data for the vertex shader. The original [MeshLine]( https://github.com/spite/THREE.MeshLine) project does this with a separate method `setGeometry`. But i wonder if we can find a way that fits better to the library conventions and is maybe more convenient.

Mugen87 committed:
MeshLine: Initial commit

Mugen87 committed:
MeshLine: More comments and better variable names

Mugen87 committed:
MeshLine: Correct typo

Mugen87 committed:
MeshLine: Support vertex colors

Mugen87 committed:
MeshLine: Clean up example

Mugen87 committed:
MeshLine: Added size attenuation

Mugen87 committed:
MeshLine: Clean up

mrdoob commented:
You're such a life saver! ❤️

mrdoob commented:
What do you think about make this transparent though? So if `THREE.Line`'s `linewidth > 1` then `WebGLRenderer` uses this code. Otherwise it goes through the usual path.

We'll have to lose `widthCallback` but, at the end of the day, I think it would be great if `THREE.Line` `linewidth` just worked.

Mugen87 commented:
>  So if THREE.Line's linewidth > 1 then WebGLRenderer uses this code. Otherwise it goes through the usual path.

After some research i was not able to find a nice integration of both approaches. But to be honest, i don't know all parts of `WebGLRenderer` in detail. If there is a smart solution to do this, i'm of course open to it. But currently it seems to me, that the code in  `WebGLRenderer` will be more complicated if we mix the rendering of line primitives and screen-space projected lines.

Therefore i would propose something different :sweat_smile:. 

`Line`, `LineLoop` and `LineSegments` should be exclusively used for rendering line primitives. Because the usage of `gl.lineWidth()` is not reliable ([reference](https://developer.mozilla.org/en-US/docs/Web/API/WebGLRenderingContext/lineWidth)), i would remove `linewidth` from `LineBasicMaterial` and `LineDashedMaterial` and render all line primitives with the default line width of `1`. If users still apply `linewidth > 1` to a `Line*` material, `three.js` could log a warning to use `MeshLine` in combination with `MeshBasicMaterial` instead.

fernandojsg commented:
> So if THREE.Line's linewidth > 1 then WebGLRenderer uses this code. Otherwise it goes through the usual path.

As this is screen space implementation it won't work properly when using WebVR. That's why I needed to create a actual geometry mesh line for [a-painter](https://aframe.io/a-painter/).
I'd planned to send a PR to add it, we could probably use three of them:
- If `non-vr` and `Linewidth == 1` -> `THREE.line`
- If `non-vr` and `LineWidth > 1` -> `THREE.MeshLineSS`
- If `vr` -> `THREE.MeshLine`.

Actually for me is also confusing the name `MeshLine` as I expected it to be real geometry mesh instead of a SS representation, so probably renaming it to a more clear name could be nice too.
Another problem I got with Jaume`s `THREE.MeshLine` implementation is that it lacks of most of the material options that we have for the rest of geometry and that could be applied to an actual geometry mesh line.

WestLangley commented on examples/webgl_meshline.html:
```diff
@@ -0,0 +1,221 @@
+
+					var object = scene.children[ i ];
+
+					if ( object instanceof THREE.Line ) {
```
`THREE.MeshLine`

Mugen87 replied:
:+1:

WestLangley commented on examples/webgl_meshline.html:
```diff
@@ -0,0 +1,221 @@
+
+				// animate line width
+
+				material1.meshLineWidth = 10 * ( 1 + 0.5 * Math.sin( 5 * time ) );
```
material2.meshLineWidth = 100 * ( 1 + 0.5 * Math.sin( 5 * time ) );

Suggestion. Vary `material2` instead, and use `OrbitControls`. It is easier to zoom and see the subtitles of what is going on.

Mugen87 replied:
Good idea!

WestLangley replied:
`100 *`
please so it is visible on small devices.

WestLangley replied:
Perhaps also set size attenuation on the first one.

What are the units of `width`?

Mugen87 replied:
Sry, overlooked the 100* thing. I've also added size attenuation to the first material.

Mugen87 replied:
'meshLineWidth' should be equal to the value of `gl.lineWidth()` so it's pixel...

WestLangley replied:
Thanks for this. You can change it later if you want. For now, it is helpful. :)

Mugen87 committed:
MeshLine: Improve Example

Mugen87 commented:
I'm not sure about the way how to handle different line implementations. We could create separate entities like `THREE.Line` or `THREE.MeshLine` and let the user decide what to use. On the other hand we can have just a single entity (e.g. `THREE.Line`) and run transparently different implementations dependent on various factors (material properties or VR/non-VR). Currently, i tend to favor the first approach. 

@fernandojsg, @WestLangley What's your opinion about this?

Mugen87 commented:
> That's why I needed to create a actual geometry mesh line for a-painter.

@fernandojsg Can you share the code for this? I am interested to see your solution :blush:

WestLangley commented:
I am not sure `MeshLine` is a good name for a screen-space approach. Trying to think of an alternate nomenclature...

- If non-vr and Linewidth == 1 -> `THREE.Line`
- If non-vr and LineWidth > 1 -> `THREE.ThickLine`  ??? or `THREE.Line` if implementation is transparent
- If vr -> `THREE.Ribbon` ???

Mugen87 committed:
MeshLine: Improve Example

mrdoob commented:
>> That's why I needed to create a actual geometry mesh line for a-painter.
>
> Can you share the code for this? I am interested to see your solution :blush:

He's probably creating a tube. Which makes me think that we could also create a tube too..

WestLangley commented:
>What are the units of width?
>>meshLineWidth should be equal to the value of gl.lineWidth() so it's pixel...

If it is consistent with THREE.Points, then fine. If not, we should figure out why. Also consider DPR.

fernandojsg commented:
Thinking about names I like @WestLangley proposal for `THREE.Ribbon` , as it's actually that, not a tube. 
I'm not sure about the approach discussed for vr/non-vr and linewidths as they're not really consistent for example regarding materials, so probably I'll just leave it to the user to choose which one they want to use.

I'll try to open a PR for the `THREE.Ribbon` later this week.

WestLangley commented:
We [used to have](https://github.com/mrdoob/three.js/issues/2523#issuecomment-9529672) `THREE.Ribbon`. I do not recall the implementation details, though. It was removed for some reason...

spite commented:
It's not a Ribbon, either, if we're being technical.

I'm afraid you're trying to cram way too functionality in Line: WebGL Lines are mathematical lines, and there's the unfortunate problem with lineWidth across platforms, which makes sense, since the definition of a line is precisely that doesn't have breadth or thickness. 

That's the main and only thing to solve for a Line replacement (and it could be built into the THREE.Line family transparently). Other stuff, like adding properties from other materials or rendering lines in VR (what does that even mean!), it should just be a Mesh, any Mesh: CurvePath, ShapePath, Ribbon, what have you.

Just my two cents.

mrdoob commented:
Agreed. I think we should leave the VR use case out of this discussion. We don't want to turn `THREE.Line` into TiltBrush. We just want to have lines with lineWidth > 1, which is something that used to work (OSX only anyway).

mrdoob commented:
> After some research i was not able to find a nice integration of both approaches. But to be honest, i don't know all parts of `WebGLRenderer` in detail. If there is a smart solution to do this, i'm of course open to it. But currently it seems to me, that the code in  `WebGLRenderer` will be more complicated if we mix the rendering of line primitives and screen-space projected lines.

So, unless someone else tries it before me, I'll use this code as reference and try to add support for `lineWidth` > 1 in `WebGLRenderer` for r86.

Mugen87 commented:
Sounds good to me. If i can manage it,  i'll have a look at the integration of the code into `WebGLRenderer`  in the next days again. For the case i make any kind of progress, i'll post it here.

fernandojsg commented:
Ok I agree, it sound good to me too, better to leave the "`ribbon`" out from this and just in case add as an external example but nothing more.
Maybe just for the `vr` case, we could just force use simple lines instead of the SS ones.

Mugen87 committed:
MeshLine: Regard DPR

Mugen87 commented:
> Also consider DPR.

@WestLangley That was a good hint! :+1:  The last commit makes the line width of `MeshLine` now consistent to `THREE.Line`.

Mugen87 committed:
MeshLine: Improve Example

Mugen87 commented:
Related issues: #10686, #10357

tomka committed:
WebGLProgram: add meshLine to list of parameter names

This is needed to distinguish WebGLProgram instances that use mesh lines
from those that don't. This is done based on a so called program code,
which is, besides other things, based on the list of known parameter
values for the parameters names from the parameter list. If two
materials only differ in whether they use mesh lines and the one that
doesn't is loaded first, then without this change no new program will be
generated if the material with enabled mesh lines is loaded afterwards,
which effectively disables mesh lines for this material.

Mugen87 committed:
Merge pull request #1 from tomka/meshline

WebGLProgram: add meshLine to list of parameter names

Mugen87 committed:
merge upstream/dev

WestLangley commented:
When the number of points is reduced, there appears to be a problem with the implementation.
```
var spline = new THREE.CatmullRomCurve3( points );
var divisions = points.length / 6; // previously points.length * 6
```
![screen shot 2017-05-10 at 2 33 07 pm](https://cloud.githubusercontent.com/assets/1000017/25915135/7cacd716-358e-11e7-86a7-703c369d32c0.png)

Mugen87 commented:
I think a too small sampling rate causes issues when calculating the line extrusions in the vertex shader. This seems to be a conceptual problem of screen-space projected lines.

Mugen87 commented:
see #11349
